feat: implementation

MQ37 opened the pull request:
This is a new integration package for LangChain that integrates Apify. This is the new recommended way instead of using the LangChain community.

This project was bootstrapped using the [langchain-cli template](https://python.langchain.com/docs/contributing/how_to/integrations/package/#boostrapping-a-new-python-package-with-langchain-cli) for integration packages.

 - Implemented LangChain tool to call Apify Actors
- Ported [ApifyDatasetLoader](https://python.langchain.com/api_reference/community/document_loaders/langchain_community.document_loaders.apify_dataset.ApifyDatasetLoader.html) from langchain community
- Ported [ApifyWrapper](https://python.langchain.com/api_reference/community/utilities/langchain_community.utilities.apify.ApifyWrapper.html) from langchain community 
- Wrote basic unit and integration tests - see `DEVELOPMENT.md` file

jirimoravcik commented on .gitignore:
Could we please extend this to e.g. https://github.com/github/gitignore/blob/main/Python.gitignore ? thank you

MQ37 replied:
thank you :+1: , extended the `.gitignore` file

jirimoravcik commented on docs/examples/tools_example.py:
```diff
@@ -0,0 +1,34 @@
+tool = ApifyActorsTool(actor_id="apify/rag-web-browser")
+
+# Use the tool directly to call the Apify Actor
+# results: list[dict] = tool.invoke(
+#    input={"run_input": {"query": "what is monero?", "maxResults": 3}}
+# )
+# for result in results:
+#    print(result)
```
Is this commented out on purpose?

MQ37 replied:
Yes, to demonstrate that the tool can be called directly

MQ37 replied:
uncommented the code so it looks as a part if the example instead of forgotten comment

jirimoravcik commented on langchain_apify/document_loaders.py:
```diff
@@ -0,0 +1,99 @@
+
+        return values
+
+    def load(self) -> list[Document]:
+        """Load documents."""
+        dataset_items = (
+            self.apify_client.dataset(self.dataset_id).list_items(clean=True).items
+        )
+        return list(map(self.dataset_mapping_function, dataset_items))
+
+    def lazy_load(self) -> Iterator[Document]:
+        """Lazy load documents."""
+        dataset_items = self.apify_client.dataset(self.dataset_id).iterate_items(
+            clean=True
+        )
+        for item in dataset_items:
+            yield self.dataset_mapping_function(item)
```
Can we describe the return values in the Docstring as well please? i.e. so people understand that lazy load yields the items instead of returning a complete list. (I know it could be inferred from the return type, but still)

MQ37 replied:
added description, thank you for suggestion

jirimoravcik commented on langchain_apify/utils.py:
```diff
@@ -0,0 +1,83 @@
+    return properties_out, required
+
+
+def create_apify_client(token: str) -> ApifyClient:
+    """Get the Apify client. Is created if not exists or token changes."""
+    if not token:
+        msg = "API token is required."
+        raise ValueError(msg)
+    client = ApifyClient(token)
+    if httpx_client := getattr(client.http_client, "httpx_client"):
+        httpx_client.headers["user-agent"] += "; Origin/langchain"
+    return client
+
+
+def create_apify_client_async(token: str) -> ApifyClientAsync:
+    if not token:
+        msg = "API token is required."
+        raise ValueError(msg)
+    client = ApifyClientAsync(token)
+    if httpx_async_client := getattr(client.http_client, "httpx_async_client"):
+        httpx_async_client.headers["user-agent"] += "; Origin/langchain"
+
+    return client
```
Can we decrease the amount of copy-paste here please?

MQ37 replied:
refactored the code, thank you

jirimoravcik commented on langchain_apify/utils.py:
```diff
@@ -0,0 +1,83 @@
+    )
+
+
+def get_actor_latest_build(client: ApifyClient, actor_id: str) -> dict:
+    """Get the latest build of an Actor from default build tag."""
+    actor = client.actor(actor_id=actor_id)
+    if not (actor_info := actor.get()):
+        msg = f"Actor {actor_id} not found."
+        raise ValueError(msg)
+
+    default_build_tag = actor_info.get("defaultRunOptions", {}).get("build")
+    latest_build_id = (
+        actor_info.get("taggedBuilds", {}).get(default_build_tag, {}).get("buildId")
+    )
+
+    if (build := client.build(latest_build_id).get()) is None:
+        msg = f"Build {latest_build_id} not found."
+        raise ValueError(msg)
+
+    return build
```
We recently introduced a new API endpoint that returns the default build, could it be used instead of this? See https://docs.apify.com/api/v2/act-build-default-get

MQ37 replied:
thank you, will use the new endpoint :+1:

jirimoravcik commented on langchain_apify/tools.py:
```diff
@@ -0,0 +1,164 @@
+        **kwargs: Any,
+    ) -> None:
+        """
+        :param actor_id: Actor name from Apify store to run.
+        :param apify_api_token: Apify API token.
```
Docstrings in this file differ from the other ones. Can we unify this please?

MQ37 replied:
unified all the docstrings, thank you

jirimoravcik reviewed (changes requested):
Hey Kuba @MQ37,
first, thanks for the pull request, it's a lot of work and it looks great.
I have some comments to improve the code:
- At Apify, we write the word "Actor" capitalized in any places where it's text (and doesn't need to follow things like `snake_case`)
- I've noticed you like to use `msg = 'error message'` and then `raise ValueError(msg)`. I personally feel like it's not necessary to separate the statements like that.
- I see that you use double quotes in your code. I believe single quotes would be better, e.g. see configuration from our Python SDK https://github.com/apify/apify-sdk-python/blob/master/pyproject.toml#L115
- I noticed some inconsistencies between how you formatted function arguments, some have the `:param` notation, some not. It would be nice to unify.
- It would be nice to have a README file that gives users more insight into what the package and Apify can do for them

Thank you

MQ37 commented on docs/examples/tools_example.py:
```diff
@@ -0,0 +1,34 @@
+from langchain_apify import ApifyActorsTool
+
+if "OPENAI_API_KEY" not in os.environ:
+    msg = "OPENAI_API_KEY environment variable is not set."
+    raise ValueError(msg)
```
using string literals can make exceptions less readable - see https://docs.astral.sh/ruff/rules/raw-string-in-exception/

jirimoravcik replied:
ok, understood.

MQ37 commented on README.md:
```diff
@@ -1,2 +1,96 @@
+
+You should configure credentials by setting the following environment variables:
+- `APIFY_API_TOKEN` - Apify API token
+- `OPENAI_API_KEY` - OpenAI API key (optional, only required if using OpenAI)
```
wanted to make this clear that they need the API key for the example snippets, but since the code sets the environment variable itself and the `docs/examples` checks if env vars are set it might not be needed - removed, thank you

MQ37 commented on README.md:
```diff
@@ -1,2 +1,96 @@
+
+## Tools
+
+`ApifyActorsTool` class exposes Apify actors from Apify. For more information, see [Apify Actors documentation](https://docs.apify.com/platform/actors).
```
added simple description, thank you for suggestion

MQ37 commented on README.md:
```diff
@@ -1,2 +1,96 @@
+
+## Document loaders
+
+`ApifyDatasetLoader` class exposes Apify datasets as document loaders. For more information, see [Apify datasets documentation](https://docs.apify.com/platform/storage/dataset).
```
also added

MQ37 commented on README.md:
```diff
@@ -1,2 +1,96 @@
+
+## Wrappers
+
+`ApifyWrapper` class wraps the Apify API to easily turn results into documents. For more information, see [Apify LangChain integration documentation](https://docs.apify.com/platform/integrations/langchain).
```
added simple description of methods available

MQ37 commented on README.md:
```diff
@@ -1,2 +1,96 @@
+- `APIFY_API_TOKEN` - Apify API token
+- `OPENAI_API_KEY` - OpenAI API key (optional, only required if using OpenAI)
+
+Learn how to get your API token in the [Apify documentation](https://docs.apify.com/platform/integrations/api).
```
improved, thank you for suggestion

MQ37 commented on README.md:
````diff
@@ -1,2 +1,96 @@
+
+`ApifyWrapper` class wraps the Apify API to easily turn results into documents. For more information, see [Apify LangChain integration documentation](https://docs.apify.com/platform/integrations/langchain).
+
+```python
+from langchain_apify import ApifyWrapper
+from langchain_core.documents import Document
+
+apify = ApifyWrapper()
+
+loader = apify.call_actor(
+    actor_id="apify/website-content-crawler",
+    run_input={
+        "startUrls": [{"url": "https://python.langchain.com/docs/get_started/introduction"}],
+        "maxCrawlPages": 10,
+        "crawlerType": "cheerio"
+    },
+    dataset_mapping_function=lambda item: Document(
+        page_content=item["text"] or "",
+        metadata={"source": item["url"]}
+    ),
+)
+documents = loader.load()
+```
````
described, thank you for suggestion

MQ37 commented on README.md:
````diff
@@ -1,2 +1,96 @@
+
+`ApifyDatasetLoader` class exposes Apify datasets as document loaders. For more information, see [Apify datasets documentation](https://docs.apify.com/platform/storage/dataset).
+
+```python
+from langchain_apify import ApifyDatasetLoader
+
+# Example dataset structure
+# [
+#     {
+#         "text": "Example text from the website.",
+#         "url": "http://example.com"
+#     },
+#     ...
+# ]
+
+loader = ApifyDatasetLoader(
+    dataset_id="your-dataset-id",
+    dataset_mapping_function=lambda dataset_item: Document(
+        page_content=dataset_item["text"],
+        metadata={"source": dataset_item["url"]}
+    ),
+)
+```
````
also added description

MQ37 commented on README.md:
````diff
@@ -1,2 +1,96 @@
+
+`ApifyActorsTool` class exposes Apify actors from Apify. For more information, see [Apify Actors documentation](https://docs.apify.com/platform/actors).
+
+```python
+import json
+from langchain_apify import ApifyActorsTool
+
+browser = ApifyActorsTool('apify/rag-web-browser')
+search_results = browser.invoke(input={
+    "run_input": {"query": "what is monero?", "maxResults": 3}
+})
+
+# use the tool with an agent
+from langchain_openai import ChatOpenAI
+from langgraph.prebuilt import create_react_agent
+os.environ["OPENAI_API_KEY"] = "YOUR_OPENAI_API_KEY"
+
+model = ChatOpenAI(model="gpt-4o-mini")
+tools = [browser]
+agent = create_react_agent(model, tools)
+
+for chunk in agent.stream(
+    {"messages": [("human", "search why is monero important?")]},
+    stream_mode="values"
+):
+    chunk["messages"][-1].pretty_print()
````
added example description

MQ37 commented on tests/unit_tests/test_document_loaders.py:
```diff
@@ -0,0 +1,47 @@
+def test_apify_dataset_loader_load() -> None:
+    with patch.object(DatasetClient, "list_items") as mock_list_items:
+        mock_list_items.return_value = ListPage(
+            data={"items": [{"text": "monero is based", "url": "https://monero.org"}]}
```
ahh, they use https://www.getmonero.org/, thank you for checking this, but changed to Apify related stuff as you suggested

MQ37 commented on tests/integration_tests/test_wrappers.py:
```diff
@@ -0,0 +1,18 @@
+from langchain_apify import ApifyWrapper
+
+
+def test_apify_wrapper_call_actor() -> None:
```
added comment, thank you for suggestion

MQ37 commented on tests/integration_tests/test_tools.py:
```diff
@@ -0,0 +1,21 @@
+
+from langchain_apify.tools import ApifyActorsTool
+
+
```
added description, thank you

MQ37 commented on langchain_apify/wrappers.py:
```diff
@@ -0,0 +1,257 @@
+
+    For details, see https://docs.apify.com/platform/integrations/langchain
+
+    Example:
+        .. code-block:: python
+
+            from langchain_apify import ApifyWrapper
+            from langchain_core.documents import Document
+
+            apify = ApifyWrapper()
+
+            loader = apify.call_actor(
+                actor_id="apify/website-content-crawler",
+                run_input={
+                    "startUrls": [
+                        {"url": "https://python.langchain.com/docs/introduction/"}
+                    ],
+                    "maxCrawlPages": 10,
+                    "crawlerType": "cheerio"
+                },
+                dataset_mapping_function=lambda item: Document(
+                    page_content=item["text"] or "",
+                    metadata={"source": item["url"]}
+                ),
+            )
+            documents = loader.load()
+    """
```
yes, this example works, thank you for noticing

MQ37 commented on tests/unit_tests/test_tools.py:
```diff
@@ -0,0 +1,56 @@
+        mock_run_actor.return_value = [{"text": "monero is based"}]
+
+        result = apify_actors_tool_fixture.invoke(
+            input={"run_input": {"query": "what is monero?", "maxResults": 3}}
```
changed monero -> apify

MQ37 commented:
> Hey Kuba @MQ37, first, thanks for the pull request, it's a lot of work and it looks great. I have some comments to improve the code:
> 
>     * At Apify, we write the word "Actor" capitalized in any places where it's text (and doesn't need to follow things like `snake_case`)
> 
>     * I've noticed you like to use `msg = 'error message'` and then `raise ValueError(msg)`. I personally feel like it's not necessary to separate the statements like that.
> 
>     * I see that you use double quotes in your code. I believe single quotes would be better, e.g. see configuration from our Python SDK https://github.com/apify/apify-sdk-python/blob/master/pyproject.toml#L115
> 
>     * I noticed some inconsistencies between how you formatted function arguments, some have the `:param` notation, some not. It would be nice to unify.
> 
>     * It would be nice to have a README file that gives users more insight into what the package and Apify can do for them
> 
> 
> Thank you

Hey Jirka @jirimoravcik, thank you for your valuable review :+1: 

- capitalized the word `Actor`, it should not appear in lowercase form anywhere in the text, thank you for noticing this
- kept `msg` and `raise` separate instead of using literal, thank you for the personal suggestion but I think ruff docs explanation makes sense - https://docs.astral.sh/ruff/rules/raw-string-in-exception/
- changed the ruff config to use single quotes and added additional rules
- fixed the inconsistencies and set google pydoc style convention, so all docstrings should follow this convention
- added README header and improved based on your suggestions, thank you (also changed the repo description)

> I'd even go further and reach out to marketing to help write a punchy paragraph here, I think it's worth it

Do you know who I can reach to help me with improving the README header?

Thank you for your time and effort :+1:

jirimoravcik commented on langchain_apify/utils.py:
```diff
@@ -0,0 +1,121 @@
+        msg = f'Failed to get the Actor object ID for {actor_id}.'
+        raise ValueError(msg)
+
+    url = f'https://api.apify.com/v2/acts/{actor_obj_id}/builds/default'
```
@danpoletaev will you be adding this to the Python client please?

danpoletaev replied:
No, it will not be added to the python/js client unless really needed. Because it will be breaking change.

Currently for `BuildClients` we use `/actor-builds` namespace instead of `/acts` namespace

danpoletaev replied:
Here is the link to zenhub task: https://app.zenhub.com/workspaces/platform-team-5f6454160d9f82000fa6733f/issues/gh/apify/apify-core/18314

"adding the /default subroute to the API client would require some major restructuring, since currently ActorClient.build() builds the Actor, it does not retrieve a specific build"

jirimoravcik commented on langchain_apify/tools.py:
```diff
@@ -0,0 +1,190 @@
+            readme = readme[:MAX_DESCRIPTION_LEN] + '...(TRUNCATED, README TOO LONG)'
+        return (
+            'Run an Apify Actor with the given input. '
+            'Here README for available Apify Actor:\n\n'
```
What does this mean? doesn't feel like a proper english sentence to me.

MQ37 replied:
Just saving on tokens :laughing: No, thank you for noticing this. went through the code and fixed typos and grammar.

jirimoravcik commented on langchain_apify/tools.py:
```diff
@@ -0,0 +1,190 @@
+
+        _description = [
+            (
+                "JSON encoded as string with input schema "
```
```suggestion
                "JSON encoded as a string with input schema "
```

jirimoravcik reviewed (approved):
Thank you

jirispilka commented on docs/examples/tools_example.py:
```diff
@@ -0,0 +1,33 @@
+import os
+
+from langchain_openai import ChatOpenAI
+from langgraph.prebuilt import create_react_agent
+
+from langchain_apify import ApifyActorsTool
+
```
why not use .env and .env.example with api keys, this would simplify it a bit

MQ37 replied:
I think since this is a package that users will usually just install and use in their project without cloning this repo I think it would not make sense to use dotenv in this example - we can set them in the example with `os.environ` instead of just checking to make it clear how to set the env vars

jirispilka commented on docs/examples/tools_example.py:
```diff
@@ -0,0 +1,33 @@
+tool = ApifyActorsTool(actor_id='apify/rag-web-browser')
+
+# Example: Use the tool directly to call the Apify Actor
+results = tool.invoke(input={'run_input': {'query': 'what is Apify?', 'maxResults': 3}})
```
```suggestion
print('Calling tool ...')
results = tool.invoke(input={'run_input': {'query': 'what are Apify Actors?', 'maxResults': 3}})
```

jirispilka commented on docs/examples/tools_example.py:
```diff
@@ -0,0 +1,33 @@
+tools = [tool]
+agent = create_react_agent(model, tools)
+
+for chunk in agent.stream(
```
The output is huuuge :), can you make it shorter?

MQ37 replied:
makes sense, could be scary to the end user - skipping all the tool messages (tool results) now and output is much shorter

jirispilka commented on langchain_apify/document_loaders.py:
```diff
@@ -0,0 +1,113 @@
+        Args:
+            values (dict): The values to validate.
+        """
+        apify_api_token = get_from_dict_or_env(
```
This can fit into one line, there are other parts of code that can be put into a single line

MQ37 replied:
removed the trailing comma, now it fits into single line

jirispilka commented on langchain_apify/tools.py:
```diff
@@ -0,0 +1,190 @@
+
+        Args:
+            run_input (Union[str, dict]): JSON input for the Actor.
+            run_manager (Optional[CallbackManagerForToolRun]): Callback manager for the
```
There is no argument run_manager

MQ37 replied:
thank you for noticing

jirispilka commented on langchain_apify/tools.py:
```diff
@@ -0,0 +1,190 @@
+        input_dict = input_dict.get('run_input', input_dict)
+        return self._run_actor(input_dict)
+
+    def _create_description(self, apify_client: ApifyClient, actor_id: str) -> str:
```
fn can be static

MQ37 replied:
thank you, I have not noticed this and ruff has this rule still in preview - https://docs.astral.sh/ruff/rules/no-self-use/

jirispilka commented on langchain_apify/tools.py:
```diff
@@ -0,0 +1,190 @@
+            f'{readme}\n\n'
+        )
+
+    def _build_tool_args_schema_model(
```
fn can be static

jirispilka commented on langchain_apify/tools.py:
```diff
@@ -0,0 +1,190 @@
+        """
+        build = get_actor_latest_build(apify_client, actor_id)
+        readme = build.get('readme', '')
+        if len(readme) > MAX_DESCRIPTION_LEN:
```
MAX_DESCRIPTION_LEN = 250 chars. This is "nothing".
Do you need to use readme then? Why not to use Actor description only and you don't need to truncate it.

MQ37 replied:
maybe just the description would be simpler and more concise - switched to description only

jirispilka commented on langchain_apify/tools.py:
```diff
@@ -0,0 +1,190 @@
+        properties, required = prune_actor_input_schema(actor_input)
+        properties = {'run_input': properties}
+
+        _description = [
```
why the description starts with `_`? I would simple use `description`, no need for extra variable

jirispilka replied:
On the second thought, what about this simplification?

Less efficient but IMO more readable - but you are not going to create 1000+ tools anyway
```
        description = f"""
               JSON encoded as a string with input schema
               (STRICTLY FOLLOW JSON FORMAT AND SCHEMA):\n\n
               f"{json.dumps(properties, separators=(',', ':'))}
               \nIF THE TOOL INPUT SCHEMA SUPPORTS IT LIMIT THE NUMBER OF RESULTS
        """
        description = description if not required else description + '\n\nRequired fields:\n' + '\n'.join(required)
  ```

MQ37 replied:
> why the description starts with `_`? I would simple use `description`, no need for extra variable

because `str` cannot be reassigned to `list[str]` type - mypy check would fail

> On the second thought, what about this simplification?

Agree with the simplification - simplified the code

jirispilka commented on langchain_apify/tools.py:
```diff
@@ -0,0 +1,190 @@
+        """Run an Apify Actor and return the output dataset.
+
+        Args:
+            actor_id: str, Actor name from Apify store to run.
+            run_input: dict, JSON input for the Actor.
+            fields: list, List of fields to extract from the dataset.
+                Other fields will be ignored.
```
```suggestion
            run_input: dict, JSON input for the Actor
```

jirispilka commented on langchain_apify/tools.py:
```diff
@@ -0,0 +1,190 @@
+            msg = 'Actor run details not found'
+            raise ValueError(msg)
+        if (run_id := details.get('id')) is None:
+            msg = 'Run id not found'
```
```suggestion
            msg = 'Run ID not found'
```

jirispilka commented on langchain_apify/tools.py:
```diff
@@ -0,0 +1,190 @@
+                run_input=run_input,
+            )
+        ) is None:
+            msg = 'Actor run details not found'
```
When you raising exception it is always good to be more specific and provide context. In this case it does not matter much but better would be:
```
f'Actor: {self._actor_id} was not started properly and details about the run were not returned'
```

jirispilka commented on langchain_apify/tools.py:
```diff
@@ -0,0 +1,190 @@
+                Other fields will be ignored.
+        """
+        if (
+            details := self._apify_client.actor(actor_id=self._actor_id).call(
+                run_input=run_input,
+            )
```
Again, this should fit into one line only

jirispilka commented on langchain_apify/utils.py:
```diff
@@ -0,0 +1,121 @@
+    return properties_out, required
+
+
+T = TypeVar('T', ApifyClient, ApifyClientAsync)
```
should it be placed on the top of file?

MQ37 replied:
since it is currently only used for the `create_apify_client` I think it makes sense to keep it just above the function - if it was used in multiple function then it would make sense to place it on the top or in separate file

jirispilka commented on langchain_apify/utils.py:
```diff
@@ -0,0 +1,121 @@
+
+    build = response.json()
+    if not isinstance(build, dict):
+        msg = f'Failed to get the latest build of the Actor {actor_id}.'
```
This is nice example of error message. Thanks!

jirispilka commented on langchain_apify/utils.py:
```diff
@@ -0,0 +1,121 @@
+        msg = f'Failed to get the Actor object ID for {actor_id}.'
+        raise ValueError(msg)
+
+    url = f'https://api.apify.com/v2/acts/{actor_obj_id}/builds/default'
```
Please let us not have URLs buried inside code. Let us move it at the top

jirispilka commented on langchain_apify/utils.py:
```diff
@@ -0,0 +1,121 @@
+T = TypeVar('T', ApifyClient, ApifyClientAsync)
+
+
+def create_apify_client(t: type[T], token: str) -> T:
```
Please do not use just variable `t`. Never.
Also. why to create typevar, what is benefit over 
```
def create_apify_client(apify_client: ApifyClient | ApifyClientAsync, token: str) -> ApifyClient | ApifyClientAsync:
```
I know, it is a bit longer, but readable

MQ37 replied:
> Please do not use just variable t. Never.

makes sense, this is unreadable for someone who did not write the code.  Renamed to `client_cls`

> Also. why to create typevar, what is benefit over

`TypeVar` is needed when you need to work with a single type in the scope. If simple union was used then you would have to always check the client type using `isinstanceof` and after that you could work with the client, otherwise linter and type checker would be angry - this was you can just pass the client type class that you want to create (sync or async) and since it is typevar type checker know that if you passed ApifyClient it has to return ApifyClient and cannot ApifyClientAsync

jirispilka commented on langchain_apify/utils.py:
```diff
@@ -0,0 +1,121 @@
+    Returns:
+        dict: The latest build of the Actor.
+    """
+    if (actor := apify_client.actor(actor_id).get()) is None:
```
```suggestion
    if not (actor := apify_client.actor(actor_id).get()):
```

jirispilka replied:
I don't think you need to test None explicitly. And the same thing below ....

jirispilka commented on langchain_apify/wrappers.py:
```diff
@@ -0,0 +1,275 @@
+            'APIFY_API_TOKEN',
+        )
+
+        client = create_apify_client(ApifyClient, apify_api_token)
+        async_client = create_apify_client(ApifyClientAsync, apify_api_token)
+
+        values['apify_client'] = client
+        values['apify_client_async'] = async_client
```
```suggestion
        values['apify_client'] = create_apify_client(ApifyClient, apify_api_token)
        values['apify_client_async'] = create_apify_client(ApifyClientAsync, apify_api_token)
```

jirispilka reviewed (changes requested):
I like it, thanks!

A couple of notes from me:
-  Github repo about: please provide text there, this is searchable field (not this Apify + 🦜🔗 = 🤖🔧🌐)
- use are missing line-length in pyproject.toml. I would suggest to use 120
```
[tool.ruff]
line-length = 120
```
- Again, some of the issues I saw should be reported by IDE. Like that you are using a variable in doc that is not in a function definition

Otherwise, good job 👍🏻

MQ37 commented on README.md:
```diff
@@ -1,2 +1,130 @@
+# use the tool with an agent
+from langchain_openai import ChatOpenAI
+from langgraph.prebuilt import create_react_agent
+os.environ["OPENAI_API_KEY"] = "YOUR_OPENAI_API_KEY"
```
it is mentioned in the Prerequisites section, but it makes sense for better UX to make it copy-paste-run - updated, thank you

MQ37 commented on README.md:
```diff
@@ -1,2 +1,130 @@
+
+For more information, see the [Apify LangChain integration documentation](https://docs.apify.com/platform/integrations/langchain).
+
+`ApifyWrapper` is useful when you need to run an Apify Actor programmatically and process the results in LangChain.
+
```
makes sense, move to the top paragraph, thank you

MQ37 commented on README.md:
````diff
@@ -1,2 +1,130 @@
+
+Example usage for `call_actor` involves running the [Website Content Crawler](https://apify.com/apify/website-content-crawler) Actor, which extracts content from webpages. The wrapper then returns the results as a list of `Document` objects containing the page content and source URL:
+```python
+from langchain_apify import ApifyWrapper
````
added the token, thank you

MQ37 commented on DEVELOPMENT.md:
```diff
@@ -0,0 +1,36 @@
+This is a development documentation for this project.
```
improved the development.md and added install instructions, thank you

MQ37 commented on README.md:
```diff
@@ -1,2 +1,130 @@
+<div align="center">
+
+<picture>
+  <img alt="Apify logo" src="/docs/logo_apify.svg" width="50%" height="50%">
```
resized to logo to 20 %

MQ37 commented:
> I like it, thanks!
> 
> A couple of notes from me:
> 
>     * Github repo about: please provide text there, this is searchable field (not this Apify + 🦜🔗 = 🤖🔧🌐)
> 
>     * use are missing line-length in pyproject.toml. I would suggest to use 120
> 
> 
> ```
> [tool.ruff]
> line-length = 120
> ```
> 
>     * Again, some of the issues I saw should be reported by IDE. Like that you are using a variable in doc that is not in a function definition
> 
> 
> Otherwise, good job 👍🏻

Thank you for your review and effort :+1: 

- refactored the code
- changed the repo description to "Apify integration for LangChain 🦜🔗 "
- set the ruff line length to 120
- found out when you run ruff in preview mode it reports these issues (for example the docstring issues and method could be static) - will check the code also in preview mode

BraniganLee commented on README.md:
```diff
@@ -1,2 +1,137 @@
+
+---
+
+This package allows you to use Apify, a platform for web scraping and data extraction, with LangChain. It provides tools to interact with Apify Actors, datasets, and API.
```
Build web scraping and automation workflows in Python by connecting Apify Actors with LangChain. This package gives you programmatic access to Apify's infrastructure - run scraping tasks, handle datasets, and use the API directly through LangChain's tools.

MQ37 replied:
Thank you Dan - updated the README :+1:

MQ37 commented on README.md:
```diff
@@ -1,2 +1,137 @@
+  <img alt="Apify logo" src="/docs/logo_apify.svg" width="20%" height="20%">
+</picture>
+
+LangChain Apify: Full-stack web scraping and data extraction platform enhanced with AI capabilities. Maintained by [Apify](https://apify.com).
```
Thank you Dan - updated the README :+1:
